### PR TITLE
Implement unlockAllFastMutexNoLock

### PR DESCRIPTION
**Pull request by schulmar**
A first try at implementing this function.
I am not sure if this achieves what the inline comment intended but it solves the deadlock/assertions (e.g. #449)  for me.

### CODE REVIEW

File: src/libdecaf/src/modules/coreinit/coreinit_fastmutex.cpp [hunk 1]
```diff
@@ -193,46 +193,59 @@ fastMutexHardUnlock(OSFastMutex *mutex)
    internal::unlockScheduler();
 }
 
+static void
+OSFastMutex_Unlock_common(OSFastMutex *mutex) {
```
Comment (brett19) [inline src/libdecaf/src/modules/coreinit/coreinit_fastmutex.cpp]: This should probably be NoLock postfixed?

File: src/libdecaf/src/modules/coreinit/coreinit_fastmutex.cpp [hunk 2]
```diff
@@ -193,46 +193,59 @@ fastMutexHardUnlock(OSFastMutex *mutex)
...
+          continue;
+       }
+
+       // Success!
+       break;
+    }
+}
+
+void
+OSFastMutex_UnlockWithoutScheduler(OSFastMutex *mutex) {
```
Comment (brett19) [inline src/libdecaf/src/modules/coreinit/coreinit_fastmutex.cpp]: NoScheduler is probably more in line withe our existing naming standards.

File: src/libdecaf/src/modules/coreinit/coreinit_fastmutex.cpp (hunk 1 above)
Comment (brett19) [reply src/libdecaf/src/modules/coreinit/coreinit_fastmutex.cpp]: Additionally, OSFastMutex_Unlock_one might be more in line with our standards, and makes it more clear what it does.

Comment (schulmar) [reply src/libdecaf/src/modules/coreinit/coreinit_fastmutex.cpp]: Well, you tell me ;) I don't know about your naming conventions and was not sure whether NoLock was directed at the mutex itself or the scheduler.

**Comment by brett19** [comment]
@exjam Can you review/merge?

**Comment by exjam** [comment]
Sorry for the slow response - I have been away.

Unfortunately your fix is not correct. NoLock implies the function is already holding the scheduler lock - and unfortunately your code would lead to the scheduler possibly being incorrectly unlocked due to unlockAllFastMutexNoLock -> OSFastMutex_UnlockWithoutScheduler -> OSFastMutex_Unlock_common -> fastMutexHardUnlock -> unlockScheduler. Also unlockAllFastMutexNoLock is not meant to lead to  testThreadCancel or reschedule called, which could happen if fastMutexHardUnlock is hit.

So basically the main problem here is that unlockAllFastMutexNoLock is actually a distinct code path from OSFastMutex_Unlock. As such it is better to just implement it in place:
https://github.com/decaf-emu/decaf-emu/commit/2c51cd4eedc77ff67fa34ef5223ec805985c8e6e

Whilst I was there I also found 2 bugs in usage of isContended:
https://github.com/decaf-emu/decaf-emu/commit/c27e76db820ce8879fadb603ca1f70f36fbafbb4

Thanks for bringing this to my attention, I was unaware that this function had been left unimplemented (I blame Brett!). For future note, reversing coreinit.rpl is a great reference for implementation details such as these. I hope this doesn't discourage you from attempting more contributions in the future :).

**Pull request closed without merging**